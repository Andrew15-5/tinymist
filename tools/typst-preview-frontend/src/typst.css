body {
  background-color: #fff;
}

.hidden {
  display: none;
}

.flex-row {
  display: flex;
  flex-direction: row;
}

.flex-row-rev {
  display: flex;
  flex-direction: row-reverse;
}

.flex-col {
  display: flex;
  flex-direction: column;
}

.disable-scrollbars::-webkit-scrollbar {
  background: transparent; /* Chrome/Safari/Webkit */
  width: 0px;
}

.disable-scrollbars {
  scrollbar-width: none; /* Firefox */
  -ms-overflow-style: none; /* IE 10+ */
}

#typst-app {
  width: fit-content;
  margin: 0;
  transform-origin: 0 0;
  background-color: #fff;
}

#typst-app.invert-colors {
  filter: invert(0.933333) hue-rotate(180deg);
}

#typst-app.invert-colors .typst-image {
  filter: invert(0) hue-rotate(0deg);
  transition: filter 0.1s ease-in-out;
}

#typst-app.invert-colors .typst-image:hover,
#typst-app.invert-colors.normal-image .typst-image {
  filter: invert(1) hue-rotate(180deg);
}

.hide-scrollbar-x {
  overflow-x: hidden;
}

.hide-scrollbar-y {
  overflow-y: hidden;
}

.typst-text {
  pointer-events: bounding-box;
  cursor: text;
}

.tsel span,
.tsel {
  left: 0;
  position: fixed;
  text-align: justify;
  white-space: pre;
  width: 100%;
  height: 100%;
  text-align-last: justify;
  color: transparent;
}

.tsel span::-moz-selection,
.tsel::-moz-selection {
  color: transparent;
  background: #7db9dea0;
}

.tsel span::selection,
.tsel::selection {
  color: transparent;
  background: #7db9dea0;
}


.pseudo-link {
  fill: transparent;
  cursor: pointer;
  pointer-events: all;
}

.image_glyph image,
.outline_glyph path,
path.outline_glyph {
  transform: matrix(1, 0, 0, 1, var(--o), 0);
}

.outline_glyph path,
path.outline_glyph {
  fill: var(--glyph_fill);
  stroke: var(--glyph_stroke);
}

.hover .typst-text {
  --glyph_fill: #66bab7;
  --glyph_stroke: #66bab7;
}

.typst-text:hover {
  --glyph_fill: #f75c2f;
  --glyph_stroke: #f75c2f;
}

.typst-dom-page {
  width: calc(var(--data-page-width, 100%) * var(--typst-dom-scale));
  height: calc(var(--data-page-height) * var(--typst-dom-scale));
  --data-text-width: calc(1px * var(--typst-dom-scale));
  --data-text-height: calc(1px * var(--typst-dom-scale));
  margin: auto;
}

.typst-html-semantics {
  position: absolute;
  z-index: 2;
  color: transparent;
  font-family: monospace;
  white-space: pre;
}

.typst-html-semantics span {
  color: transparent;
  font-family: monospace;
  transform-origin: left top;
  position: absolute;
  display: inline-block;
  left: 0;
  top: 0;
}

.typst-content-hint {
  position: absolute;
  display: inline-block;
  width: 1px;
  height: 1px;
  overflow: hidden;
}

.typst-html-semantics a {
  position: absolute;
  display: inline-block;
}

/* set transparent itself */
.typst-content-group {
  pointer-events: visible;
}

.typst-html-semantics span::-moz-selection {
  color: transparent;
  background: #7db9dea0;
}

.typst-html-semantics span::selection {
  color: transparent;
  background: #7db9dea0;
}

.typst-html-semantics *::-moz-selection {
  color: transparent;
  background: transparent;
}

.typst-html-semantics *::selection {
  color: transparent;
  background: transparent;
}

.typst-content-fallback {
  color: transparent;
  background: transparent;
}

.pseudo-link,
.typst-text {
  pointer-events: none;
}


.typst-jump-ripple,
.typst-debug-react-ripple {
  width: 0;
  height: 0;
  background-color: transparent;
  position: absolute;
  border-radius: 50%;
}

.typst-jump-ripple {
  border: 2px solid #66bab7;
}

.typst-debug-react-ripple {
  border: 1px solid #cb1b45;
}

@keyframes typst-jump-ripple-effect {
  to {
    width: 10vw;
    height: 10vw;
    opacity: 0.01;
    margin: -5vw;
  }
}

@keyframes typst-debug-react-ripple-effect {
  to {
    width: 3vw;
    height: 3vw;
    opacity: 0.1;
    margin: -1.5vw;
  }
}

.typst-svg-cursor {
  /* https://www.elevenforum.com/t/change-text-cursor-blink-rate-in-windows-11.12409/ */
  /* 1.2 seconds per blink is the default value on gtk and the slowest value on windows */
  animation: 1.2s blink step-start infinite;
}

@keyframes blink {
  50% {
    opacity: 0;
  }
}
