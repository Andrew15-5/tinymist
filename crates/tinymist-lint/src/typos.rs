//! `typos` as spell checking backend.

use typos::tokens::{Identifier, Word};
use typos_cli::policy;
use typst::{
    diag::EcoString,
    syntax::{
        ast::{self, AstNode},
        Source, Span, SyntaxKind, SyntaxNode,
    },
};

#[derive(Debug, Clone, Copy)]
enum TextPos {
    Markup,
    Code,
}

fn scan_idents(src: &Source, mut f: impl FnMut(TextPos, &str, Span)) {
    fn scan_group(src: &SyntaxNode, f: &mut impl FnMut(TextPos, &str, Span)) {
        for ch in src.children() {
            scan_idents_impl(ch, f)
        }
    }

    fn scan_text(src: &SyntaxNode, f: &mut impl FnMut(TextPos, &str, Span)) {
        let t = src.text();
        if !t.is_empty() {
            f(TextPos::Markup, t, src.span());
        }
    }

    fn scan_ident(src: &SyntaxNode, f: &mut impl FnMut(TextPos, &str, Span)) {
        let t = src.text();
        if !t.is_empty() {
            f(TextPos::Code, t, src.span());
        }
    }

    fn scan_idents_impl(src: &SyntaxNode, f: &mut impl FnMut(TextPos, &str, Span)) {
        if matches!(src.kind(), SyntaxKind::Error) {
            return;
        }

        use SyntaxKind::*;
        match src.kind() {
            Markup => scan_group(src, f),
            Text => scan_text(src, f),
            Space => {}
            Linebreak => {}
            Parbreak => {}
            Escape => {}
            Shorthand => {}
            SmartQuote => {}
            Strong => scan_group(src, f),
            Emph => scan_group(src, f),
            Raw => {}
            RawLang => {}
            RawDelim => {}
            RawTrimmed => {}
            Link => {}
            Label => {}
            Ref => {}
            RefMarker => {}
            Heading => scan_group(src, f),
            HeadingMarker => {}
            ListItem => scan_group(src, f),
            ListMarker => {}
            EnumItem => scan_group(src, f),
            EnumMarker => {}
            TermItem => scan_group(src, f),
            TermMarker => {}
            Equation => scan_group(src, f),
            Math => scan_group(src, f),
            MathAlignPoint => scan_group(src, f),
            MathDelimited => scan_group(src, f),
            MathAttach => scan_group(src, f),
            MathPrimes => scan_group(src, f),
            MathFrac => scan_group(src, f),
            MathRoot => scan_group(src, f),
            Hash | LeftBrace | RightBrace | LeftBracket | RightBracket | LeftParen | RightParen
            | Comma | Semicolon | Colon | Star | Underscore | Dollar | Plus | Minus | Slash
            | Hat | Prime | Dot | Eq | EqEq | ExclEq | Lt | LtEq | Gt | GtEq | PlusEq | HyphEq
            | StarEq | SlashEq | Dots | Arrow | Root | Not | And | Or | None | Auto | Let | Set
            | Show | Context | If | Else | For | In | While | Break | Continue | Return
            | Import | Include | As | Bool | Int | Float | Numeric | Str => {}
            MathIdent | Ident => scan_ident(src, f),
            Code => scan_group(src, f),
            CodeBlock => scan_group(src, f),
            ContentBlock => scan_group(src, f),
            Parenthesized => scan_group(src, f),
            Array => scan_group(src, f),
            Dict => scan_group(src, f),
            Named => scan_group(src, f),
            Keyed => scan_group(src, f),
            Unary => scan_group(src, f),
            Binary => scan_group(src, f),
            FieldAccess => {}
            FuncCall => {
                let fc = src.cast::<ast::FuncCall>().unwrap();
                scan_idents_impl(fc.args().to_untyped(), f);
            }
            Args => {
                let args = src.cast::<ast::Args>().unwrap();
                for arg in args.items() {
                    match arg {
                        ast::Arg::Pos(p) => scan_idents_impl(p.to_untyped(), f),
                        ast::Arg::Named(p) => scan_idents_impl(p.expr().to_untyped(), f),
                        ast::Arg::Spread(..) => {}
                    }
                }
            }
            Params => {
                let args = src.cast::<ast::Params>().unwrap();
                for arg in args.children() {
                    match arg {
                        ast::Param::Pos(p) => scan_idents_impl(p.to_untyped(), f),
                        ast::Param::Named(p) => scan_idents_impl(p.expr().to_untyped(), f),
                        ast::Param::Spread(..) => {}
                    }
                }
            }
            Spread => scan_group(src, f),
            Closure => scan_group(src, f),
            LetBinding => scan_group(src, f),
            SetRule => scan_group(src, f),
            ShowRule => scan_group(src, f),
            Contextual => scan_group(src, f),
            Conditional => scan_group(src, f),
            WhileLoop => scan_group(src, f),
            ForLoop => scan_group(src, f),
            ModuleImport => {}
            ImportItems => {}
            RenamedImportItem => {}
            ModuleInclude => {}
            LoopBreak => scan_group(src, f),
            LoopContinue => scan_group(src, f),
            FuncReturn => scan_group(src, f),
            Destructuring => scan_group(src, f),
            DestructAssignment => scan_group(src, f),
            LineComment => {}
            BlockComment => {}
            Error => {}
            Eof => {}
        }
    }

    scan_idents_impl(src.root(), &mut f)
}

/// Check for typos in the source.
pub fn typos_check(src: &Source, mut f: impl FnMut(Span, usize, Vec<EcoString>)) {
    let default_policy = policy::Policy::default();

    let dict = default_policy.dict;

    let mut report = |span: Span, offset: usize, cc: Status<'_>| match cc {
        Status::Valid => {}
        Status::Invalid => {
            f(span, offset, vec![]);
        }
        Status::Corrections(corrections) => {
            let cc = corrections.into_iter().map(|c| c.as_ref().into()).collect();
            f(span, offset, cc);
        }
    };

    use typos::tokens::Case;
    use typos::Status;
    scan_idents(src, |pos, t, span| match pos {
        TextPos::Markup => {
            println!("checking {:?} -> {pos:?}:{t:?}", src.range(span).unwrap());
            let ident = Word::new_unchecked(t, Case::Lower, 0);
            if let Some(st) = dict.correct_word(ident) {
                report(span, 0, st);
            }
        }
        TextPos::Code => {
            println!("checking {:?} -> {pos:?}:{t:?}", src.range(span).unwrap());
            let ident = Identifier::new_unchecked(t, Case::None, 0);
            let res = dict.correct_ident(ident);
            match res {
                Some(Status::Valid) => {}
                Some(st @ (Status::Invalid | Status::Corrections(..))) => report(span, 0, st),
                None => {
                    for ident in ident.split() {
                        if let Some(st) = dict.correct_word(ident) {
                            report(span, 0, st);
                        }
                    }
                }
            }
        }
    });
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::*;

    use typst::syntax::Source;

    fn check_typos(src: &Source) -> String {
        let mut res = String::new();
        typos_check(src, |span, _, cc| {
            res.push_str(&format!("{:?} -> {:?}\n", src.range(span).unwrap(), cc));
        });

        res
    }

    #[test]
    fn test() {
        snapshot_testing("typos", &|src| {
            assert_snapshot!(check_typos(&src));
        });
    }
}

// /// Suggestion for change in a text.
// #[derive(Debug, Clone)]
// pub struct Suggestion {
//     pub source: String,
//     pub message: String,
//     pub span: Option<MappedSpan>,
//     pub replacements: Vec<String>,
// }

// pub fn nlp_check_docs(doc: Arc<TypstDocument>) -> Option<Vec<Suggestion>> {
//     let annotated = super::text_export::TextExporter::default()
//         .annotate(doc)
//         .ok()?;
//     let suggestions = nlp_check(&annotated.content);
//     let spans = suggestions
//         .iter()
//         .map(|s| s.span().char().clone())
//         .collect::<Vec<_>>();
//     let spans = annotated.map_back_spans(spans);
//     Some(
//         suggestions
//             .into_iter()
//             .zip(spans)
//             .map(|(suggestion, span)| Suggestion {
//                 source: suggestion.source().to_string(),
//                 message: suggestion.message().to_string(),
//                 span,
//                 replacements: suggestion
//                     .replacements()
//                     .iter()
//                     .map(|x| x.to_string())
//                     .collect(),
//             })
//             .collect(),
//     )
// }

// pub fn diag_from_suggestion(suggestion: Suggestion) ->
// typst::diag::SourceDiagnostic {     typst::diag::SourceDiagnostic {
//         severity: typst::diag::Severity::Warning,
//         message: eco_format!("{:?}", suggestion.message),
//         span: suggestion
//             .span
//             .map(|s| s.span.span)
//             .unwrap_or(Span::detached()),
//         trace: Default::default(),
//         hints: Default::default(),
//     }
// }
