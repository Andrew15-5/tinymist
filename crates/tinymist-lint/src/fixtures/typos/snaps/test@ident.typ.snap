---
source: crates/tinymist-lint/src/typos.rs
expression: check_typos(&src)
input_file: crates/tinymist-lint/src/fixtures/typos/ident.typ
---

